mlflow/utils/autologging_utils/safety.py

```diff
@@ -773,8 +780,50 @@ def _validate_autologging_run(autologging_integration, run_id):
     ), "Autologging run with id {} has a non-terminal status '{}'".format(run_id, run.info.status)
 
 
+def __is_arg_exempt_from_validation(
+    autologging_integration,
+    function_name,
+    arg,
+    pos_arg_index=None,
+    kwarg_name=None,
+):
+    """
+    This function is responsible for determining whether or not an argument is exempt from autolog
+    safety validations. This includes both type checking and immutable checking.
+
+    WARNING: Exemptions should NOT be introduced unless absolutely necessary. If deemed necessary,
+             clear reasons must be provided as comment in addition to thorough integration tests.
+
+    :param autologging_integration: the name of the autologging integration
+    :param function_name: the name of the function that is being validated
+    :param arg: the actual argument
+    :param pos_arg_index: the index of the argument, if it is a positional argument
+    :param kwarg_name: the name of the argument, if it is a keyword argument
+    :return: True or False
+    """
+
+    # When extracting implicitly defined `batch_size` in the case that `x` is a generator or a
+    # generator class, we need to consume and restore the first element back to the generator to
+    # calculate the `batch_size`. This means that:
+    # 1. The type of `x` will become 'generator' regardless if user provided `x` as a generator or a
+    #    custom generator class.
+    # 2. The instance of `x` will be different, since we reconstructed the generator after consuming
+    #    the first element.
+    if autologging_integration == mlflow.tensorflow.FLAVOR_NAME:
+        if function_name == "fit":
+            if pos_arg_index == 1 or kwarg_name == "x":
+                from mlflow.utils.autologging_utils import is_generator
+
+                return is_generator(arg)
```

Comment:
For testing purposes, can we move this to a more generic allowlist form like we had in the previous commits? I propose a list of named tuples with the following attributes:

1. autologging integration
2. function name
3. argument name
4. argument index
5. type_fn (a function that takes an object and returns True if the object matches the argument type, False otherwise)